pkg/daemon/vm_controller.go

```diff
@@ -75,80 +90,230 @@ func (r *VMReconciler) reconcile(ctx context.Context, vm *virtv1alpha1.VirtualMa
-						r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedResume", "Failed to resume VM")
+
+					vm.Status.Migration.TargetPort = port
+					vm.Status.Migration.TargetNodeIP = r.NodeIP
+					vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationTargetReady
+				}
+			case virtv1alpha1.VirtualMachineMigrationTargetReady:
+				if vm.Status.NodeName == r.NodeName {
+					ctx, cancel := context.WithCancel(context.Background())
+					migrationControlBlock.SendDomainCancelFunc = cancel
+
+					tlsConfig := &tls.Config{
+						InsecureSkipVerify: true,
+						GetClientCertificate: func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+							return tlsutil.LoadCert(daemonCertDirPath)
+						},
+					}
+					if err := r.RelaySocketToTCP(ctx, filepath.Join(getVMSocketDirPath(vm), "send-migration.sock"), fmt.Sprintf("%s:%d", vm.Status.Migration.TargetNodeIP, vm.Status.Migration.TargetPort), tlsConfig); err != nil {
+						return fmt.Errorf("start source relay: %s", err)
+					}
+
+					sendMigrationErrChan := make(chan error, 1)
+					go func() {
+						if err := r.getCloudHypervisorClient(vm).VmSendMigration(ctx, &cloudhypervisor.SendMigrationData{
+							DestinationUrl: "unix:/var/run/virtink/send-migration.sock",
+						}); err != nil {
+							sendMigrationErrChan <- err
+						}
+					}()
+					migrationControlBlock.SendDomainErrCh = sendMigrationErrChan
+					vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationRunning
+				}
+			case virtv1alpha1.VirtualMachineMigrationRunning:
+				if vm.Status.NodeName == r.NodeName {
+					var vmPod corev1.Pod
+					vmPodKey := types.NamespacedName{
+						Name:      vm.Status.VMPodName,
+						Namespace: vm.Namespace,
+					}
+					if err := r.Get(ctx, vmPodKey, &vmPod); err != nil {
+						return fmt.Errorf("get VM Pod: %s", err)
+					}
+
+					if vmPod.Status.Phase == corev1.PodSucceeded {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationSent
+					} else if migrationControlBlock.SendDomainErrCh == nil {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
 					} else {
-						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Resumed", "Resumed VM")
+						select {
+						case err := <-migrationControlBlock.SendDomainErrCh:
+							if err != nil {
+								r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedMigrate", "Failed to migrate VM to %s: %s", vm.Status.Migration.TargetNodeName, err)
+								vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
+							}
+						default:
+							return errors.New("VM is migrating")
+						}
+					}
+					if sendDomainCancelFunc := migrationControlBlock.SendDomainCancelFunc; sendDomainCancelFunc != nil {
+						sendDomainCancelFunc()
 					}
-				default:
-					// ignored
 				}
+			case virtv1alpha1.VirtualMachineMigrationSent:
+				if vm.Status.Migration.TargetNodeName == r.NodeName {
+					// TODO handle VmNotCreated error
```

Comment:
Solve this TODO?

Reply:
we don't know the VM is creating or it will never be created(maybe something wrong). raise the error might be a better choice now.

and now we just raise an error during VM is migrating instead of return a `ctrl.Result` with requeue. maybe we need another PR to fix all these questions.

Reply:
We can capture certain type of errors in the main reconcile func and convert it to a requeue. But yeah, it can be another commit